### PR TITLE
Endgame bionic power source: Microreactor CBM

### PR DESCRIPTION
**Pull request by Lorith**
Adds a new bionic power source for endgame characters - a stripped down minireactor.  As would be expected, nuclear power results in power over time, and has the side effect of radiation.  The reactor can be upgraded to cleanse its own radiation, as well as have higher efficiency, it continues to run at all times unless out of fuel, and will cause pain/damage if it pushes out more energy than you can hold.  An upgrade kit can be crafted, allowing you to upgrade your internal reactor without removing it, and removing a reactor rapidly converts all consumed plutonium into radiation.  The fuel for the reactor is more efficient than plutonium cells in a battery, but quite a bit lower than in an actual minireactor in a vehicle.

Additionally, a large reactor takes a lot of room.  This is incompatible with internal storage and the internal furnace.

Overall, this results in a very powerful generation system that requires continual attention to ensure it does not go haywire and kill you, which it will gladly do if you put too much fuel into it, or forget about the radiation.

### CODE REVIEW

**Comment by Zireael07** [comment]
Cool!

**Comment by Lain-** [comment]
Yea this is pretty nice. I like it.

**Comment by Coolthulhu** [comment]
It's a cool idea, in a wacky sci-fi "plutoniumpunk" style. About as orky as a cyborg running on diesel.
A few issues:

* The whole battery overloading thing is simply a bad idea. It makes no sense for advanced electronics not to detect this and step down production and it only adds needless micromanagement to the game
* Why does plutonium filter have a passive effect? It makes it harder to limit power consumption (say, when you want to use the solar panels), plus makes the above battery overloading problem worse
* Metal tank wouldn't fit in a person's chest. Little metal tank would make more sense in the recipes.
* When trying to install, the message that "you don't have enough space in chest" is not informative. Should mention the offending CBMs explicitly

I'm not too keen on upgrading CBMs (especially when it increases their speed by a factor of 10 and doubles efficiency) and I don't like the the whole tank->reactor transfer thing and balancing things with it, but those are just my opinions and not objective problems.

**Comment by Lorith** [comment]
Honestly, I hadn't checked the tank size, good point on that.

The battery overloading is based on the idea of the reactor being too large/heavy to begin with and everything possible being stripped out, including all control software and most of the shielding.  It also makes it require more attention than the others which are just insert A, gain power, at worst getting a little addiction.  The way I see it, the higher gains need a higher risk if you screw up.  Used in moderation, it won't get to the point of causing damage.  But if someone goes overboard with power, they need to make sure they don't overload.

As for the passive effect...  I know there was a reason, but I honestly can't figure it out logically anymore, so will be removing that.

For the install message, I was unsure the maximum message size, so was trying to keep it small.  Does it auto-format to fit the screen, or is there some other way to make sure it is not too large?

And the upgrade does not increase the efficiency, though it does increase the speed significantly and consumes some of the power so it does not irradiate you.  The tank > reactor transfer is mostly to control the amount of power generated over time so it starts small, increases, then decreases again as fuel runs out.  I want to avoid it being a instant-power which is what happened without that - you would get a few ticks of massive power regen followed by a small trickle.  The same unit of fuel will produce the same power, except that the advanced reactor sometimes provides less power to avoid radiation.  The only change is in the speed it generates and the maximum limit per tick.  The high end of the generation on either reactor is not reached unless you use far more fuel than is reasonably safe, though it is useful for those with huge amounts of power storage.

**Comment by Coolthulhu** [comment]
Balancing power generation based on player's ability to store and utilize it is still not good. Why would someone with active hydraulic muscles be safer from the harmful effects of the reactor? Same for nanobots and similar effects that can be invoked without any time cost.
Plus, it's not like the new cyborgs (with low power storage) need an extra nerf.

If anything, damage should happen when too much power is generated regardless of whether it gets stored or wasted.

I misread that part about efficiency, yeah.

I'm trying to comprehend the advanced reactor radiation avoidance.
In the example of maximal usage (500 energy - is it per turn? If so, nothing uses that much, except stacking repair nanobots), it gets 250 slow rad, then steps down twice to generate only 50 slow rad, consume 500 plutonium units and produce 498 energy. Am I reading it right?

In this case, when generating 199 energy, it would generate 99 slow rad (maximum value to avoid stepdown). This would result in at most 1 radiation per 5 turns. Compare to iodine effect: 1 radiation cured per about 16 turns.

**Comment by Lorith** [comment]
Good point on the radiation.  Currently it looks like I screwed up on the reactor part, though it also adds to the radiation over time based on the contents of the storage tank and reactor fuel, so it is possible to add several points of radiation per turn depending on how much fuel you have added.  The way I am thinking of is to keep the radiation cleaning roughly equal to the same cost as running the radiation scrubber manually, possibly a bit higher, simply to reduce the effort with the advanced version, though it does appear I screwed up on my math and it is only using 1/2 of what it should right now.  The slow radiation builds up over time before adding to normal radiation, and the reactor can only clean it before it becomes normal radiation.  Does the radiation per power multiplier look good for the normal reactor?  Can't think of why I have the adv different besides the auto-cleanser, so will be changing that.

As for the limits on energy, it is per turn and designed for the extreme-lategame cyborgs to be able to recharge in a time period not measured in days.  Additionally, it should not be too much of a problem, as it is intended for lategame, being craft-only from a rare item, requiring extremely high skills, and the low-level fuel only provides 250 energy.  One of those alone should not produce enough energy to cause any negative side effects besides radiation.  Pain only occurs from 10 power more than your limit per turn, damage only from 50 over.  If you keep it as a slow charge, it will not do more than sting a bit if you add fuel while full of energy.

Personally, I am somewhat attached to the idea of it doing damage from too much output, though I am considering adding in damage when providing high rates of power as well.  I just feel that it should not be a "shove everything possible in with no danger" like it would without the overcharge damage.

So current plans: metal tank > little metal tank, remove passive filter effect, better messages for install failure, redo radiation for the adv reactor plus increase cleanse cost.

Am I missing any of the points you made besides the overload?  And thanks for the feedback, I hope it doesn't seem like I am trying to disagree with everything, I just have a few things I really like the idea of with it and want to keep.

**Comment by Coolthulhu** [comment]
Damage from too much output isn't a bad idea in itself, but it should not be limited by energy capacity. It is easy to keep energy down - you can just spam say, blood filter, radiation scrubber, repair nanobots - even repeatedly extending/retracting claws would work, because most bionics take 0 time to use.

If you want to keep overload damage, tie it to something not dependent on power storage. For those with full energy, wasting plutonium is enough of a punishment. Plutonium isn't cheap.

On normal reactor it looks like it generates 1 radiation per 250 energy produced. This is a bit low - you don't take damage until 150 radiation, so here you could safely get almost 40k energy before any damage starts, even without iodine or scrubbing.

I'd say scrap slow radiation completely. Just make it a random chance - it works like this with all other sources of radiation. Actually, with pretty much all fractions (fuel usage). Since radiation builds up over time, this should statistically have the same effect as slow rad accumulation, but with less numbers to track.

File: src/bionics.cpp [hunk 1]
```diff
@@ -1277,6 +1283,24 @@ bool player::install_bionics(const itype &type)
         popup("invalid / unknown bionic id %s", bioid.c_str());
         return false;
     }
+    if (bioid == "bio_reactor" || bioid == "bio_advreactor") {
+		if (has_bionic("bio_furnace") || has_bionic("bio_storage") || has_bionic("bio_reactor") || has_bionic("bio_advreactor")) {
+			popup(_("You lack the required space in your torso!"));
```
Comment (KA101) [inline src/bionics.cpp]: Little more descriptive: your X is getting in the way.  We don't track bionic "slots" so folks will justifiably wonder why they lack the space.

File: src/player.cpp [hunk 2]
```diff
@@ -7640,6 +7643,90 @@ void player::suffer()
...
+                    } else {
+                        power_gen = reactor_plut * 0.025;
+                    }
+                    slow_rad += (power_gen * 2); //lack of radiation shielding means that you get bonus radiation
+                }
+                reactor_plut -= power_gen;
+                if (power_gen > (max_power_level - power_level)) {
+                    power_gen -= (max_power_level - power_level);
+                    power_level = max_power_level;
+                    if (power_gen >= 50) { //more power than you can store is painful and harmful
```
Comment (KA101) [inline src/player.cpp]: Should be more along the lines of power gen exceeding a certain amount per turn (haven't looked, and this'd need the other systems to be cross-checked), you overload your internal circuits and take damage/stun from that, IMO.

File: src/player.h [hunk 3]
```diff
@@ -920,6 +920,7 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
         int power_level, max_power_level;
         int hunger, thirst, fatigue;
         int stomach_food, stomach_water;
+		int tank_plut, reactor_plut, slow_rad;
```
Comment (KA101) [inline src/player.h]: eek, tabs.  Check your text editor: most can set the tab key to insert spaces instead, so indents are uniform between editors.

**Comment by Lorith** [comment]
That was certainly odd.  I had been editing with tab set to put in spaces, so no idea how that happened.  I found the option to show whitespace though due to you pointing it out, and checked everything I touched.  Lots of tabs replaced with spaces all over now.

File: src/player.cpp [hunk 4]
```diff
@@ -1876,16 +1879,16 @@ void player::memorial( std::ofstream &memorial_file, std::string epitaph )
...
         int distance_from_city = abs(g->cur_om->dist_from_city(cur_loc));
         if(distance_from_city > nearest_city.s + 4) {
-            //~ First parameter is a pronoun (“He”/“She”), second parameter is a terrain name.
+            //~ First parameter is a pronoun (�He�/�She�), second parameter is a terrain name.
             kill_place = string_format(_("%s was killed in a %s in the wilderness."),
                          pronoun.c_str(), tername.c_str());
 
         } else if(distance_from_city >= nearest_city.s) {
-            //~ First parameter is a pronoun (“He”/“She”), second parameter is a terrain name, third parameter is a city name.
+            //~ First parameter is a pronoun (�He�/�She�), second parameter is a terrain name, third parameter is a city name.
```
Comment (BevapDin) [inline src/player.cpp]: Your editor is not set to use UTF-8 and changed those UTF-8 sequence to something else. Make sure you load and save all files of this project with the UTF-8 encoding.

**Comment by Lorith** [comment]
Think that should fix the accidental changes, and also fixed a smaller oops with spacing.

**Pull request merged**